src/block/block_compact.c

```diff
@@ -67,6 +69,203 @@ __wt_block_compact_get_progress_stats(WT_SESSION_IMPL *session, WT_BM *bm,
+ */
+static void
+__block_compact_estimate_remaining_work(WT_SESSION_IMPL *session, WT_BLOCK *block)
+{
+    WT_EXT *ext;
+    wt_off_t avg_block_size, compact_start_off, extra_space, file_size, last, write_off;
+    uint64_t n, pages_to_move, total_pages_to_move;
+    int compact_pct_tenths;
+    bool skip;
+
+    /*
+     * We must have relocated at least some interesting number of pages for any estimates below to
+     * be worthwhile.
+     */
+    if (block->compact_pages_rewritten < WT_THOUSAND)
+        return;
+
+    /* Assume that we have already checked whether this file can be skipped. */
+    WT_ASSERT(session, block->compact_pct_tenths > 0);
+
+    /*
+     * Get the average block size that we encountered so far during compaction. Note that we are not
+     * currently accounting for overflow pages, as compact does not currently account for them
+     * either.
+     */
+    avg_block_size = (wt_off_t)WT_ALIGN(
+      block->compact_bytes_rewritten / block->compact_pages_rewritten, block->allocsize);
+
+    __wt_verbose_debug2(session, WT_VERB_COMPACT,
+      "%s: the average block size is %" PRId64 " bytes (based on %" PRIu64 " relocations)",
+      block->name, avg_block_size, block->compact_pages_rewritten);
+    __wt_verbose_debug2(session, WT_VERB_COMPACT,
+      "%s: will move blocks from the last %d%% of the file", block->name,
+      block->compact_pct_tenths * 10);
+
+    /*
+     * We would like to estimate how much data will be moved during compaction, so that we can
+     * inform users how far along we are in the process. We will estimate how many pages are in the
+     * last part of the file (typically the last 10%) and where they will be written using the
+     * first-fit allocation, and then repeat as long as we continue to make progress, to emulate the
+     * behavior of the actual compaction. This does not account for all complexities that we may
+     * encounter, but the hope is that the estimate would be still good enough.
+     */
+
+    __wt_spin_lock(session, &block->live_lock);
+
+    compact_pct_tenths = block->compact_pct_tenths;
+    extra_space = 0;
+    file_size = block->size;
+    pages_to_move = 0;
+    total_pages_to_move = 0;
+    write_off = 0;
+
+    for (;;) {
+        compact_start_off = (10 - compact_pct_tenths) * file_size / 10;
+        if (write_off >= compact_start_off)
+            break;
+
+        /* Estimate how many pages we would like to move, just making use the live checkpoint. */
+        last = compact_start_off;
+        WT_EXT_FOREACH (ext, block->live.avail.off) {
+            if (ext->off < compact_start_off)
+                continue;
+            if (ext->off > file_size)
+                break;
+            if (ext->off > last) {
+                pages_to_move += (uint64_t)((ext->off - last) / avg_block_size);
+                last = ext->off + ext->size;
+            }
+        }
+        pages_to_move += (uint64_t)((file_size - last) / avg_block_size);
```

Comment:
I did it this way specifically because of rounding, so that we don't overestimate the number of leaf pages too much—especially because we already overestimate that by not accounting for internal pages in this estimate. I am not sure how much this would matter in practice though, but it doesn't seem like this optimization is worth it.